quinn-proto/src/range_set/array_range_set.rs

```diff
@@ -0,0 +1,301 @@
+///   a single cache-friendly contiguous range.
+///
+/// `ArrayRangeSet` is especially useful for tracking ACK ranges where the amount
+/// of ranges is usually very low (since ACK numbers are in consecutive fashion
+/// unless reordering happens).
+#[derive(Debug, Default)]
+pub struct ArrayRangeSet(TinyVec<[Range<u64>; ARRAY_RANGE_SET_INLINE_CAPACITY]>);
+
+/// The capacity of elements directly stored in [`ArrayRangeSet`]
+///
+/// An inline capacity of 2 is chosen to keep `SentFrame` below 128 bytes.
+const ARRAY_RANGE_SET_INLINE_CAPACITY: usize = 2;
+
+/// An iterator over ranges in an [`ArrayRangeSet`]
+pub struct ArrayRangeSetIter<'a> {
+    inner: &'a TinyVec<[Range<u64>; ARRAY_RANGE_SET_INLINE_CAPACITY]>,
+    front_idx: usize,
+    back_idx: usize,
+}
+
+impl<'a> Iterator for ArrayRangeSetIter<'a> {
+    type Item = Range<u64>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.front_idx == self.back_idx {
+            return None;
+        }
+
+        let elem = self.inner.get(self.front_idx).cloned();
+        self.front_idx += 1;
+        elem
+    }
+}
+
+impl<'a> DoubleEndedIterator for ArrayRangeSetIter<'a> {
+    fn next_back(&mut self) -> Option<Self::Item> {
+        if self.back_idx == self.front_idx {
+            return None;
+        }
+
+        self.back_idx -= 1;
+        self.inner.get(self.back_idx).cloned()
+    }
+}
+
+/// An iterator over elements in an [`ArrayRangeSet`]
+pub struct ArrayRangeSetEltIter<'a> {
+    inner: ArrayRangeSetIter<'a>,
+    start: u64,
+    end: u64,
+}
+
+impl<'a> Iterator for ArrayRangeSetEltIter<'a> {
+    type Item = u64;
+    fn next(&mut self) -> Option<u64> {
+        if self.start == self.end {
+            let next_range = self.inner.next()?;
+            self.start = next_range.start;
+            self.end = next_range.end;
+        }
+        let x = self.start;
+        self.start += 1;
+        Some(x)
+    }
+}
+
+impl Clone for ArrayRangeSet {
+    fn clone(&self) -> Self {
+        // tinyvec keeps the heap representation after clones.
+        // We rather prefer the inline representation for clones if possible,
+        // since clones (e.g. for storage in `SentFrames`) are rarely mutated
+        if self.0.is_inline() || self.0.len() > ARRAY_RANGE_SET_INLINE_CAPACITY {
+            return Self(self.0.clone());
+        }
+
+        let mut vec = TinyVec::new();
+        vec.extend_from_slice(self.0.as_slice());
+        Self(vec)
+    }
+}
+
+impl ArrayRangeSet {
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    pub fn iter(&self) -> ArrayRangeSetIter<'_> {
+        ArrayRangeSetIter {
+            inner: &self.0,
+            front_idx: 0,
+            back_idx: self.0.len(),
+        }
+    }
+
+    pub fn elts(&self) -> ArrayRangeSetEltIter<'_> {
+        ArrayRangeSetEltIter {
+            inner: self.iter(),
+            start: 0,
+            end: 0,
+        }
+    }
```

Comment:
Do we need these types to be named? We could save a decent chunk of code with `impl (DoubleEnded)Iterator` and combinators here. Even if we do need named types, newtyping the combinators (`Copied<slice::Iter>`, `Flatten<ArrayRangeSetIter>`) might be more concise than writing out the logic by hand.

Reply:
I can replace them with `impl Iterator`

Reply:
The intended upside to using `impl Iterator` is that we can replace the handwritten `Iterator` implementations with combinators, which will be much more concise.